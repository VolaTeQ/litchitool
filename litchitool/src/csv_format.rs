use std::io::Read;

pub use csv;
use csv::Reader;
use tracing::Level;

use crate::{
    error::LitchiError,
    mission::{
        Action, AltitudeMode, Coordinate, GimbalPitchMode, LitchiMission, MissionConfig,
        PhotoInterval, Waypoint, POI,
    },
};

macro_rules! parse_chunk {
    ($record:expr => $($name:ident ($type:ty) $idx:expr),+) => {
        $(
            let $name: $type = $record.get($idx).ok_or(LitchiError::CsvMissingField($idx))?.parse()?;
        )+
    }
}

pub fn read_from_csv<R: Read>(mut reader: Reader<R>) -> Result<LitchiMission, LitchiError> {
    let mut waypoints: Vec<Waypoint> = vec![];
    let mut pois: Vec<POI> = vec![];

    for (record_index, record) in reader.records().enumerate() {
        let row_span = tracing::span!(Level::TRACE, "Parsing row of CSV", row = record_index);
        let _enter = row_span.enter();
        let record = record?;
        if record.len() != 46 {
            Err(LitchiError::IncorrectRecordLength(record.len(), 46))?;
        }

        const ACTIONS_OFFSET: usize = 8;
        const ACTIONS_COUNT: usize = 15;
        const ACTIONS_END: usize = ACTIONS_OFFSET + ACTIONS_COUNT * 2;

        parse_chunk!(record =>
            latitude             (f64) 0,
            longitude            (f64) 1,
            altitude             (f32) 2,
            heading              (f32) 3,
            curve_size           (f32) 4,
            rotation_dir         (i32) 5,
            gimbal_mode          (i32) 6,
            gimbal_pitch_angle   (i32) 7,
            altitude_mode       (i16) ACTIONS_END,
            speed               (f32) ACTIONS_END + 1,
            poi_latitude        (f64) ACTIONS_END + 2,
            poi_longitude       (f64) ACTIONS_END + 3,
            poi_altitude        (f32) ACTIONS_END + 4,
            poi_altitude_mode   (i16) ACTIONS_END + 5,
            photo_time_interval (f32) ACTIONS_END + 6,
            photo_distance_interval (f32) ACTIONS_END + 7
        );

        let mut heading = heading;

        let gimbal_mode = GimbalPitchMode::try_from(gimbal_mode)
            .map_err(|err| LitchiError::TryFromPrimitiveError(err.number.to_string()))?;
        let altitude_mode = AltitudeMode::try_from(altitude_mode)
            .map_err(|err| LitchiError::TryFromPrimitiveError(err.number.to_string()))?;
        let photo_time_interval = Some(photo_time_interval).filter(|interval| *interval > 0.);
        let photo_distance_interval =
            Some(photo_distance_interval).filter(|interval| *interval > 0.);

        let actions = (0..ACTIONS_COUNT)
            .map(|action_i| -> Result<Option<Action>, LitchiError> {
                parse_chunk!(record =>
                    action_type (i32) ACTIONS_OFFSET + action_i * 2,
                    action_param (i32) ACTIONS_OFFSET + 1 + action_i * 2
                );

                Ok(match action_type {
                    -1 => None,
                    0 => Some(Action::StayFor(action_param as f32 / 1000.)),
                    1 => Some(Action::TakePhoto),
                    2 => Some(Action::StartRecording),
                    3 => Some(Action::StopRecording),
                    4 => Some(Action::RotateAircraft(action_param)),
                    5 => Some(Action::TiltCamera(action_param)),
                    n => Err(LitchiError::InvalidActionType(n))?,
                })
            })
            .filter_map(|res| match res {
                Ok(Some(action)) => Some(Ok(action)),
                Ok(None) => None,
                Err(error) => Some(Err(error)),
            })
            .collect::<Result<Vec<_>, _>>()?;

        let poi = {
            let poi_altitude_mode = AltitudeMode::try_from(poi_altitude_mode)
                .map_err(|err| LitchiError::TryFromPrimitiveError(err.number.to_string()))?;

            if poi_longitude != 0. && poi_latitude != 0. && poi_altitude != 0. {
                Some(POI {
                    coordinate: Coordinate(poi_latitude, poi_longitude),
                    altitude: poi_altitude,
                    altitude_mode: poi_altitude_mode,
                })
            } else {
                None
            }
        };

        let coordinates = Coordinate(latitude, longitude);

        if let Some(poi) = &poi {
            heading = coordinates.heading_towards(&poi.coordinate) as f32;
        }

        let poi_index = poi.map(|poi| {
            pois.iter()
                .position(|search_poi| &poi == search_poi)
                .unwrap_or_else(|| {
                    pois.push(poi);
                    pois.len() - 2
                })
        });

        waypoints.push(Waypoint {
            coordinate: coordinates,
            altitude,
            altitude_mode,
            heading,
            curve_size,
            gimbal_mode,
            gimbal_pitch_angle,
            speed,
            poi_index,
            actions,
            turn_mode: 0,
            photo_interval: photo_time_interval
                .map(PhotoInterval::Time)
                .or_else(|| photo_distance_interval.map(PhotoInterval::Distance)),
            repeat_actions: 1,
            rotation_dir,
            stay_time: 3,
            max_reach_time: 0,
        })
    }

    LitchiMission::new(waypoints, pois, MissionConfig::default())
}
