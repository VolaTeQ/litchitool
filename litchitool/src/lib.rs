pub mod csv_format;
pub mod error;
pub mod mission;

#[cfg(test)]
mod tests {
    use crate::csv_format;

    #[test]
    fn test_convert_mission() {
        const TEST_MISSION_CSV: &[u8] = include_bytes!("../test/litchi_mission.csv");

        let mission = csv_format::read_from_csv(csv::Reader::from_reader(TEST_MISSION_CSV))
            .expect("Could not parse test mission from csv");

        let binary = mission.to_binary();

        insta::assert_debug_snapshot!(binary);
    }

    // TODO: More tests
}
