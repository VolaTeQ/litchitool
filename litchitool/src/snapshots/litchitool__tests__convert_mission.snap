---
source: litchitool/src/lib.rs
expression: binary
---
b"lchm\0\0\0\x02\0\0\0\x01\0\0\0\0A\0\0\0A`\0\0\0\0\0\x01\0\x0b\0\0\0\0\0\0\0\0\0\0\0\0\0\x0fA\xf0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x03\0\0@@\xcf0\xb92 z@\x17\xac\x8c\xb3N\x1a9>L\xcc\xcd\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x01A\xf0\0\0\0\0\0\0\0\0\0\0@&ff\0\x03\0\0@>\t\x07\x87YU\xb0\xc0\x0e\xb2\xe6\x99c\xcb\x8eDz\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x01A\xf0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x03\0\0@B\xe0\xeb\xbb\xac\x87L\xc02\x13\\\xd3,yrDz\0\0\0\0\0\0\0\0\0\0\0\0\0\x01\0\0\0\x01\0\0\0\x01\0\0\0\0A\xf0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x03\0\0@Fh\x02\x80\x1a\xb3Q\xc0(\xe6\xb9\xa6X\xf2\xe3Dz\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x01A\xf0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x03\0\0@F\xb7c<\xd3\x9a\xfd\xc0\x12\xddsL\xb1\xe5\xc7Dz\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x01A\xf0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x03\0\0@<\xcfy\x04\xbcd\xd6@\x19\xc8\x8c\xb3N\x1a9Dz\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x01A\xf0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x03\0\0@3p\xaa\xce`\xe9\x8d?\xe7te\x9ap\xd1\xc8Dz\0\0\0\0\0\0\0\0\0\0\0\0\0\x01\0\0\0\x01\0\0\0\x04\0\0\0ZA\xf0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x03\0\0@4DJ\xb9\xc2\xf06\xc02V\xdc\xd3,yrDz\0\0\0\0\0\x02\0\0\0\n\0\0\0\x02\0\0\0\x01\0\0\0\x05\0\0\0\0\0\0\0\0\0\x01\x86\xa0A\xf0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x03\0\0@C\\\x9f\ro'\x17@\x15\x90\x8c\xb3N\x1a9Dz\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x01C\xde\0\0\0\0\0\0C\n\0\0\0\0\0\0\0\x03\0\0@F\xbb\xed\xe5\x8d\x1d\x8d@.\xe6\x03\xa4\x89\xa7^Dz\0\0\0\0\0\0\0\0\0\0\0\0\0\x01\0\0\0\x01\0\0\0\0\0\0\x0b\xb8A\xf0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x03\0\0@Kq\rxm\xf4\x86@\x14P\x07I\x13N\xbcDz\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x01A\xf0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x03\0\0@I\t)Z\xd4\x0e\x1b\xc0#\x8c\xfc[vX\x97Dz\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x01A\xf0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x03\0\0@H\x0e\xbb\xf3\x93\x97\x99\xc04h\xfe-\xbb,KDz\0\0\0\0\0\0\0\0\0\0\0\0\0\x01\0\0\0\x01\0\0\0\x02\0\0\0\0A\xf0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x03\0\0@D \x9cN\x16\xcbo\xc07O~-\xbb,KDz\0\0\0\0\0\0\0\0\0\0\0\0\0\x01\0\0\0\x01\0\0\0\0\0\0\0\0C\xde\0\0\0\0\0\0C\n\0\0\0\0\0\0\0\x03\0\0@Cs2\xb2YK\x8b\xc0A\x04?\x16\xdd\x96&>L\xcc\xcd\0\0\0\0\0\0\0\0\0\0\0\x01\0\0\0\x01\0\0\0\x03\0\0\0\0\0\0\0\0\0\0A\xf0\0\0\xff\xff\xff\xff\0\0A\xf0\0\0\xff\xff\xff\xff\0\0A\xf0\0\0\xff\xff\xff\xff\0\0A\xf0\0\0\xff\xff\xff\xff\0\0A\xf0\0\0\xff\xff\xff\xff\0\0A\xf0\0\0\xff\xff\xff\xff\0\0A\xf0\0\0\xff\xff\xff\xff\0\0A\xf0\0\0\xff\xff\xff\xff\0\0A\xf0\0\0\xff\xff\xff\xff\0\0C\xde\0\0\xff\xff\xff\xff\0\0A\xf0\0\0\xff\xff\xff\xff\0\0A\xf0\0\0\xff\xff\xff\xff\0\0A\xf0\0\0\xff\xff\xff\xff\0\0A\xf0\0\0\xff\xff\xff\xff\0\0C\xde\0\0\xff\xff\xff\xff\0\0\0\x08\0\0\0\x08\0\0\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0A@\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0A \0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0\xbf\x80\0\0"
